lldb/source/Plugins/Process/mach-core/RegisterContextUnifiedCore.cpp

```diff
@@ -0,0 +1,300 @@
+      const RegisterSet *old_set = additional_reginfo_up->GetRegisterSet(idx);
+      for (size_t jdx = 0; jdx < m_register_sets.size(); jdx++) {
+        if (strcmp(m_register_sets[jdx].name, old_set->name) == 0) {
+          metadata_regset_to_combined_regset[idx] = jdx;
+          found_match = true;
+          break;
+        }
+      }
+      // This metadata RegisterSet is a new one.
+      // Add it to the combined RegisterSet array.
+      if (!found_match) {
+        RegisterSet new_set;
+        copy_regset_name(new_set, *old_set);
+        metadata_regset_to_combined_regset[idx] = m_register_sets.size();
+        m_register_sets.push_back(new_set);
+      }
+    }
+  }
+
+  // Set up our RegisterInfo array, one RegisterSet at a time.
+  // The metadata registers may be declared to be in a core thread
+  // register set (e.g. "General Purpose Registers"), so we scan
+  // both core registers and metadata registers when creating the
+  // combined RegisterInfo arrays.
+  for (size_t combined_regset_idx = 0;
+       combined_regset_idx < m_register_sets.size(); combined_regset_idx++) {
+    uint32_t registers_this_regset = 0;
+    if (combined_regset_idx < core_thread_regctx_sp->GetRegisterSetCount()) {
+      const RegisterSet *regset =
+          core_thread_regctx_sp->GetRegisterSet(combined_regset_idx);
+      for (size_t j = 0; j < regset->num_registers; j++) {
+        uint32_t reg_idx = regset->registers[j];
+        const RegisterInfo *reginfo =
+            core_thread_regctx_sp->GetRegisterInfoAtIndex(reg_idx);
+        RegisterValue val;
+        if (!reginfo->value_regs &&
+            core_thread_regctx_sp->ReadRegister(reginfo, val)) {
+          m_regset_regnum_collection[combined_regset_idx].push_back(
+              m_register_infos.size());
+          m_register_infos.push_back(*reginfo);
+          registers_this_regset++;
+        }
+      }
+    }
+    if (additional_reginfo_up) {
+      // Find the register set in the metadata that matches this register
+      // set, then copy all its RegisterInfos.
+      for (size_t setidx = 0;
+           setidx < additional_reginfo_up->GetNumRegisterSets(); setidx++) {
+        if (metadata_regset_to_combined_regset[setidx] == combined_regset_idx) {
+          const RegisterSet *regset =
+              additional_reginfo_up->GetRegisterSet(setidx);
+          for (size_t j = 0; j < regset->num_registers; j++) {
+            uint32_t reg_idx = regset->registers[j];
+            const RegisterInfo *reginfo =
+                additional_reginfo_up->GetRegisterInfoAtIndex(reg_idx);
+            m_regset_regnum_collection[combined_regset_idx].push_back(
+                m_register_infos.size());
+            // register names in DynamicRegisterInfo are ConstString stored;
+            // we can reuse the char* pointers here without retaining the
+            // DynamicRegisterInfo.
+            m_register_infos.push_back(*reginfo);
+            registers_this_regset++;
+          }
```

Comment:
Isn't this code identical to the codeon line 212? Can you extract a helper/lambda that takes a `const RegisterSet *`?

Reply:
Yeah lol, there's like 3 lines I could pull out of these two loops into a lambda.

Reply:
yeah punting on this, it's more lines overall to put those in a lambda.